File: JAVA/786. K-th Smallest Prime Fraction.java
class Solution {
    public int[] kthSmallestPrimeFraction(int[] arr, int k) {
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> Double.compare(b[0], a[0]));
        for (int i = 0; i < arr.length; i++) {
            pq.offer(new double[] {
                -1.0 * arr[i] / arr[arr.length - 1],  
                i,  
                arr.length - 1 
            });
        }
        while (--k > 0) {
            double[] cur = pq.poll();
            int numeratorIndex = (int) cur[1];
            int denominatorIndex = (int) cur[2] - 1;
            if (denominatorIndex > numeratorIndex) {
                pq.offer(new double[] {
                    -1.0 * arr[numeratorIndex] / arr[denominatorIndex],  
                    numeratorIndex,  
                    denominatorIndex  
                });
            }
        }
        double[] result = pq.poll();
        return new int[]{arr[(int) result[1]], arr[(int) result[2]]};
    }
}

File: C++/786. K-th Smallest Prime Fraction.cpp
class Solution {
public:
    vector<int> kthSmallestPrimeFraction(vector<int>& arr, int k) {
        int n = size(arr); 
        priority_queue<pair<double, pair<int, int>>, vector<pair<double, pair<int, int>>>, greater<>> pq; 
        for (int i = 0; i < n-1; ++i) 
            pq.push({(double) arr[i]/arr[n-1], {i, n-1}}); 
        
        int i = 0, j = 0; 
        while (k--) {
            auto elem = pq.top(); pq.pop(); 
            i = elem.second.first;
            j = elem.second.second; 
            if (i < j-1) pq.push({(double) arr[i]/arr[j-1], {i, j-1}}); 
        }
        return {arr[i], arr[j]}; 
    }
};
